airflow/decorators/sensor.py

```diff
@@ -0,0 +1,93 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from inspect import signature
+from typing import Any, Callable, Dict, Optional, Tuple
+
+from airflow.decorators.base import get_unique_task_id, task_decorator_factory
+from airflow.sensors.base import BaseSensorOperator
+
+
+class DecoratedSensorOperator(BaseSensorOperator):
+    """
+    Wraps a Python callable and captures args/kwargs when called for execution.
+
+    :param python_callable: A reference to an object that is callable
+    :type python_callable: python callable
+    :param task_id: task Id
+    :type task_id: str
+    :param op_kwargs: a dictionary of keyword arguments that will get unpacked
+        in your function (templated)
+    :type op_kwargs: dict
+    :param op_args: a list of positional arguments that will get unpacked when
+        calling your callable (templated)
+    :type op_args: list
+    :param kwargs_to_upstream: For certain operators, we might need to upstream certain arguments
+        that would otherwise be absorbed by the DecoratedOperator (for example python_callable for the
+        PythonOperator). This gives a user the option to upstream kwargs as needed.
+    :type kwargs_to_upstream: dict
+    """
+
+    template_fields = ('op_args', 'op_kwargs')
+    template_fields_renderers = {"op_args": "py", "op_kwargs": "py"}
+
+    # since we won't mutate the arguments, we should just do the shallow copy
+    # there are some cases we can't deepcopy the objects (e.g protobuf).
+    shallow_copy_attrs = ('python_callable',)
+
+    def __init__(
+        self,
+        *,
+        python_callable: Callable,
+        task_id: str,
+        op_args: Tuple[Any],
+        op_kwargs: Dict[str, Any],
+        **kwargs,
+    ) -> None:
+        kwargs.pop('multiple_outputs')
+        kwargs['task_id'] = get_unique_task_id(task_id, kwargs.get('dag'), kwargs.get('task_group'))
+        self.python_callable = python_callable
+        # Check that arguments can be binded
+        signature(python_callable).bind(*op_args, **op_kwargs)
+        self.op_args = op_args
+        self.op_kwargs = op_kwargs
+        super().__init__(**kwargs)
+
+    def poke(self, context: Dict) -> bool:
+        return self.python_callable(*self.op_args, **self.op_kwargs)
+
+
+def sensor(python_callable: Optional[Callable] = None, multiple_outputs: Optional[bool] = None, **kwargs):
+    """
+    Wraps a function into an Airflow operator.
+
+    Accepts kwargs for operator kwarg. Can be reused in a single DAG.
+
+    :param python_callable: Function to decorate
+    :type python_callable: Optional[Callable]
+    :param multiple_outputs: if set, function return value will be
+        unrolled to multiple XCom values. List/Tuples will unroll to xcom values
+        with index as key. Dict will unroll to xcom values with keys as XCom keys.
+        Defaults to False.
```

Comment:
```suggestion
    :param multiple_outputs: If True, the function return value will be unrolled to multiple XCom values.
        Dict will unroll to XCom values with keys as XCom keys. Defaults to False.
```
Tuples and Lists are not currently supported for `multiple_outputs`.

Reply:
Thanks! I removed "multiple_outputs" from the key word arg list since sensor operator doesn't have outputs.

Reply:
@josh-fell re

Tuples and Lists are not currently supported for multiple_outputs.

are you sure? how come it indicates it is supported in the docstring e.g. for _PythonDecoratedOperator?

Reply:
Separately....

It seems to me that it would be nice if we could optionally return an XCom value with python sensor

so the return type would be.... and this is a mouthful but... `Optional[Union[bool, Tuple[bool, Any]]]`

or something like  that

so, if you return bool it's as normal

and when sensing is successful, you can return `True, {"some": "information"}`, i.e. a 2-tuple where the first element is the poke boolean and the second element is the xcom  return.

Reply:
update: [draft PR](https://github.com/apache/airflow/pull/20547) for allowing return of xcom from poke method.  with this in place we could allow return of XCom (even with multiple outputs) with this decorator.

Reply:
As discussed in https://github.com/apache/airflow/pull/20546#issuecomment-1002476070  - I believe it's not going to work if we return Tuple  (because the tuple is always "truthy" and it will work very unexpectedly if the sensors are used with old Airflow version - they will never wait)

But (as proposed there) i think we might get it work (and in backwards-compatible way)  if we change the contract of poke method and allow "truthy" values to be retuned not only Bool and we automatically push non-Bool, Truthy return values to Xcom - see the comment for details.

Reply:
> @josh-fell re
> 
> Tuples and Lists are not currently supported for multiple_outputs.
> 
> are you sure? how come it indicates it is supported in the docstring e.g. for _PythonDecoratedOperator?

@dstandish 
Yeah, originally as part of AIP-31 there was proposed support for unrolling Tuples and Lists. However, there seemed to be a consensus that handling these datatypes added more complexity than value (see [this comment](https://github.com/apache/airflow/pull/8962/files#r428948073) and [this one](https://github.com/apache/airflow/pull/10349#issuecomment-680918458) when `multiple_outputs` via type inference was added). The original docstrings still contained the verbiage about Lists/Tuples and has remained.

When handling outputs in `DecoratedOperator`, there is an [explicit check](https://github.com/apache/airflow/blob/8a03a505e1df0f9de276038c5509135ac569a667/airflow/decorators/base.py#L137-L160)  on the return value confirming it is a dict type. Also the type-hint inference only handles dict types.

Despite this, #15813 was logged earlier this year to at least add Tuples to type-hint inference of `multiple_outputs`. Perhaps this should be revisited?

The docstrings are planned to be corrected as part of #19608.

Reply:
> As discussed in [#20546 (comment)](https://github.com/apache/airflow/pull/20546#issuecomment-1002476070) - I believe it's not going to work if we return Tuple (because the tuple is always "truthy" and it will work very unexpectedly if the sensors are used with old Airflow version - they will never wait)

Yeah I agree that that PR as is would be problematic and would need changes before merge.  I look forward to collaborating on figuring out the best way to do this.  I think we should make it happen before we introduce this taskflow sensor because depending on our choices we could be (and might have to be) stricter with the contract re return types in taskflow sensor and don't want to have to change it after release.

I did have a concern with your truthy-based return namely pandas dataframes don't behave, and  commented on that pr.

There's a third  way demonstrated in https://github.com/apache/airflow/pull/20547 which introduces a PokeReturnValue class which is an optional return value which can resolve this ambiguity problem.  We could have the taskflow sensor behave in such a way that tuple or list is converted to a PokeReturnValue instance and then the base sensor can know what to do with it.

Lemme know your thoughts.

Reply:
Hi @dstandish - I see your point of https://github.com/apache/airflow/pull/20547 and like the idea of returning a PokeReturnValue type. One approach I can think of to make it backward compatible is to update the return type of poke() function as "union[bool, PokeReturnValue]", and override "__bool__()" in PokeReturnValue. The other approach is to introduce a new type of sensor, ex "RichSensor" and make it return a PokeReturnValue type.

Reply:
@potiuk is right - we cannot have poke() return PokeReturnValue without breaking backward compatibility. The alternative I can think of is to add a new type of Sensor and let the poke() method return "PokeReturnValue". I'd like to see better suggestions.

Reply:
while i'd prefer to avoid a special class, i just want to point out that using something like PokeReturnValue would not be backward incompatible. 

all sensors that do not return PokeReturnValue would continue to work just as they do.  the issue Jarek highlights is, i think, a different one. namely, that a sensor which _did_ make use of PokeReturnValue would not be usable on older versions of airflow.  which makes sense cus it's a new feature.  

but the concern i think is then, to release that sensor, do you also need to bump the min airflow version of the provider.  maybe there would be a way around that. Namely, maybe it would be ok to include that (new) sensor in the provider, not bump min airflow, and just let only that sensor not be usable with older airflow versions -- perhaps raising a hellpful error message in that scenario.

separately, since we're only talking about _new_ sensors (since old ones would continue to work exactly as they do), it would also be possible to design it in a backward compatible way.  e.g. check airrflow version and use the appropriate return type.  _However_, i suspect very strongly that, with such a sensor, the xcom behavior would be integral to its operation, so there would be no point in enabling this kind of compatibility with old airflow versions (since they would not support it anyway!). all of which leads me to conclude that this is maybe a non-issue.

in any case, there was some discussion of the matter, and opinion seems to be leaning against adding xcom support in sensor poke method.  i plan to catchup on the latest discussions tomorrow.  though support for your PR is good! just trying to sort out quickly whether we ought also to take this opportunity to support xcom too.

thanks

Reply:
Yeah. Also I think there is a possibility that we might be more relaxed about "airlfow 2.1/2.2 compatibility". We seem to have more and more accumulated changes in Airlfow tht might make Airflow 2.3 another release that we can bump all our providers as min-version. There are some very good reasons why we might want to do it for all providers for Airflow 2.3 and this opens up the really interesting possibility of making contract changes (and PokeReturnValue in this case might be back in game). 

I started discussion here in the devlist: https://lists.apache.org/thread/csczm7xmnntdz9wtjbod8pqgt13zoggo 

and if we decide that one of the next waves of community providers should be Airflow 2.3+ only -  we might actually consider some more "bold" change here.

Reply:
@dstandish @potiuk Thanks for the clarification! I look forward to the discussion. I'd be interested in working on supporting xcom with the sensor's poke API if we decide to go this route.

Reply:
HI @dstandish, @potiuk - I have raised a PR https://github.com/apache/airflow/pull/20656 to capture your idea of supporting XCom return values for the base sensor operator. Please have a look if you have time. Thanks!